### PR TITLE
[ROCm] Add per-call decode budget to sparse-MLA indexer

### PR DESCRIPTION
**Pull request by reger-men**
## Purpose

The ROCm AITER sparse-MLA indexer decode path allocates a fresh fp32 logits tensor on every call to `rocm_fp8_paged_mqa_logits`: `(heads, B*next_n, max_model_len)` on the stage1 3D path, or `(B*next_n, max_model_len)` on the gfx942/gfx950 fused 2D path. With long context and a large running batch the per-call working set is multi-GiB; across ~60 layers per decode forward the caching allocator fragments and OOMs. Prefill had the same problem and got `VLLM_SPARSE_INDEXER_MAX_LOGITS_MB` plus chunking in #36178. This is the decode mirror.

Two pieces, decode-only, ROCm-only:

- the fp32 logits buffer is now drawn from a per-process workspace keyed by (device, shape) and refilled with `-inf` between calls, so no per-layer alloc/free of multi-GiB tensors;
- `VLLM_SPARSE_INDEXER_DECODE_MAX_MB` (default 4096) bounds the per-call working set. When a single call would exceed the budget the batch is split into sub-batches that fit and copied into the full output. The aiter kernel does not consume `schedule_metadata` on ROCm, so the metadata doesn't need to be re-derived per chunk. Set to 0 to restore the legacy unbounded behaviour.

Prefill is not touched. The CUDA path is not touched (this file only runs on ROCm via the AITER sparse-MLA backend). Both `_ON_GFX942` and `_ON_GFX950` are covered, plus the stage1 3D else-branch that older arches still hit.

Related: #36178 is the prefill-side mirror (same shape of fix, opposite forward-pass phase). #41002 is in flight in the same area, happy to rebase on top of it after it merges.

## Test Plan

1. Numeric equivalence test added at `tests/v1/attention/test_rocm_aiter_mla_sparse_decode.py`. Skipped on non-ROCm and when the AITER `paged_mqa_logits` module is not importable, so it sits with the rest of the ROCm-only kernel tests. On a ROCm gate it asserts the chunked path returns bit-identical logits to the full-batch call.

   ```bash
   pytest -s -v tests/v1/attention/test_rocm_aiter_mla_sparse_decode.py
   ```

2. Synthetic micro-bench on a single MI355X (gfx950), 60 back-to-back calls into `rocm_fp8_paged_mqa_logits` with a pre-allocated paged kv at `B=1024, next_n=1, heads=64, head_dim=128, block_size=64, max_model_len=131072`. Stress shape, picked so the 3D per-call working set is well above any reasonable budget. Measured peak `torch.cuda.max_memory_allocated()` minus the paged-kv baseline, sweeping `VLLM_SPARSE_INDEXER_DECODE_MAX_MB` against the legacy code.

3. End-to-end re-run of the GLM-5.1-FP8 customer repro that motivated this:

   ```bash
   vllm bench serve --random-input-len 1 --random-output-len 2 --num-prompts 1024 \
       --ignore-eos --temperature 0 --seed 0 \
       --gpu-memory-utilization 0.9 -tp 8 --enforce-eager
   ```

## Test Result

Indexer working set at the high-water mark, 60-layer loop:

| setting                                            | indexer working set @ peak |
| -------------------------------------------------- | -------------------------: |
| legacy (no patch)                                  |                  33792 MiB |
| `VLLM_SPARSE_INDEXER_DECODE_MAX_MB` unset (= 4096) |                  ~4160 MiB |
| `VLLM_SPARSE_INDEXER_DECODE_MAX_MB=2048`           |                  ~2080 MiB |
| `VLLM_SPARSE_INDEXER_DECODE_MAX_MB=1024`           |                  ~1040 MiB |

The 33.8 GiB legacy delta is exactly `heads * B * next_n * max_model_len * 4` bytes for the 3D stage1 buffer; chunking brings it down to `chunk_b * heads * next_n * max_model_len * 4` plus a one-shot 2D accumulator. Wall time across all four settings stays between 1.09 and 1.24 s for the 60-call loop, the only overhead is one `copy_` per chunk into the persistent output buffer.

End-to-end on GLM-5.1-FP8, the decode-time HIP OOM that originally fired at ~24 GiB per layer no longer occurs; 1017/1024 prompts succeed, the 7 remaining failures are bench-client side ("never received a valid chunk") and clear with `--max-concurrency 256`.

### CODE REVIEW

**Comment by reger-men** [comment]
Thanks for the review. Pushed 0f37431:

- 2D helper: round `rows` up to the next pow2 for the cache key, fill only `[:rows]`, return the contiguous leading-dim slice. Cache size now bounded to O(log2(max_num_seqs * next_n)) entries per (device, cols).
- 3D helper: same pow2 rounding. The workspace stays sized to `alloc_rows` because `ws[:, :rows, :]` is non-contiguous and the AITER `deepgemm_fp8_paged_mqa_logits_stage1` kernel requires contiguous output. The kernel only writes `[:, :rows, :]` (it iterates over `q_fp8`'s batch dim, not the workspace), so the result is sliced after the `sum(dim=0)`. The chunked 3D path slices the same way before the `copy_`.
- Race condition: didn't add a buffer pool. Callers run on the per-rank vLLM v1 forward thread, so a shape-colliding concurrent fill can't happen in normal use. The comment block above the helpers now calls this out and notes that any multi-threaded caller has to allocate its own buffer.
- Added `test_decode_workspace_rounding` covering the helper bookkeeping.

**Comment by reger-men** [comment]
Rebased onto current vllm main to clear the conflict. One thing to flag while reviewing: resolving it meant taking the newer fused-decode sanitize, the pre-kernel out_logits.fill_(-inf) is gone and it now relies on the post-kernel nan_to_num_(-inf) from #44527/#49714. Our per-call budget + batch chunking sits on top and sanitizes the assembled output once after the loop. @tjtanaa @AndreasKaratzas mind a review pass? It also needs the ready/verified label before CI runs, which I can't add (0 merged PRs).

**Comment by reger-men** [comment]
@tjtanaa @AndreasKaratzas I rebased this onto current main, so the merge conflict is cleared and the PR is mergeable again (head b10eef112). The only red now is pre-run-check, so it just needs the ready label to run CI plus a review pass.

**Comment by AndreasKaratzas** [comment]
Can you check if the new test file is still needed? We have some new aiter mla tests introduced over the last 7 days.

**Comment by reger-men** [comment]
I checked, and the test still covers something the new ones dont. The aiter MLA tests added this past week are on the attention backend, not the sparse indexer: `test_rocm_aiter_mla_fp8_decode_routing.py` (#51011) checks the Gluon vs asm decode routing and head padding, and `test_rocm_aiter_mla_mtp_split.py` checks the MTP metadata builder. Neither one calls `rocm_fp8_paged_mqa_logits` or the decode budget.

The only test that touches `rocm_fp8_paged_mqa_logits` is `test_paged_mqa_logits_do_not_contain_nan` in `test_rocm_triton_attn_dsv4.py`. It fakes the kernel to fill NaN on a 1x1 shape and asserts the output is NaN-free, so it only exercises the single-call path and the sanitize. It never sets `VLLM_SPARSE_INDEXER_DECODE_MAX_MB` and never enters the chunk loop that this PR adds.

My test is the only one that forces the per-call split (budget=1) and asserts the chunked output is bit-identical to the full-batch call, which is the behavior this PR introduces. The only overlap is the budget=0 arm, which reruns the full path that the NaN test already covers, so I can drop that arm and keep just the chunked-vs-full check if you'd prefer it trimmed.

**Comment by reger-men** [comment]
@tjtanaa @AndreasKaratzas this is rebased and mergeable, the only red is the ready/verified label gate. Could you add the label so the suite runs, and review?
